prefetcher/EIP/EIP.cc

```diff
@@ -0,0 +1,871 @@
+  uint64_t time_i = l1i_hist_table_head_time[l1i_cpu_id];
+  uint64_t req_time = 0;
+  uint32_t num_skipped = 0;
+  for (uint32_t count = 0, i = first; count < L1I_HIST_TABLE_ENTRIES; count++, i = (i + L1I_HIST_TABLE_MASK) % L1I_HIST_TABLE_ENTRIES) {
+    // Against the time overflow
+    if (req_time == 0 && l1i_hist_table[l1i_cpu_id][i].tag == tag && time_i + latency >= current_core_cycle[l1i_cpu_id]) { // Its me (miss or late prefetcher)
+      req_time = time_i;
+    } else if (req_time) { // Not me (check only older than me)
+      if (l1i_hist_table[l1i_cpu_id][i].tag == tag) {
+        return 0; // Second time it appeared (it was evicted in between) or many for the same set. No entangle
+      }
+      if (time_i + latency <= req_time && l1i_hist_table[l1i_cpu_id][i].tag) {
+        if (skip == num_skipped) {
+          return l1i_hist_table[l1i_cpu_id][i].tag;
+        } else {
+          num_skipped++;
+        }
+      }
+    }
+    time_i -= l1i_hist_table[l1i_cpu_id][i].time_diff;
+  }
+  return 0;
+}
+
+// TIMING TABLES
+
+#define L1I_SET_BITS 6
+#define L1I_TIMING_MSHR_SIZE (L1I_PQ_SIZE + L1I_MSHR_SIZE + 2)
+#define L1I_TIMING_MSHR_TAG_BITS 42
+#define L1I_TIMING_MSHR_TAG_MASK (((uint64_t)1 << L1I_HIST_TAG_BITS) - 1)
+#define L1I_TIMING_CACHE_TAG_BITS (L1I_TIMING_MSHR_TAG_BITS - L1I_SET_BITS)
+#define L1I_TIMING_CACHE_TAG_MASK (((uint64_t)1 << L1I_HIST_TAG_BITS) - 1)
+
+// We do not have access to the MSHR, so we aproximate it using this structure
+typedef struct __l1i_timing_mshr_entry {
+  bool valid;              // 1 bit
+  uint64_t tag;            // L1I_TIMING_MSHR_TAG_BITS bits
+  uint64_t bere_line_addr; // 58 bits
+  uint64_t timestamp;      // L1I_TIME_BITS bits // time when issued
+  bool accessed;           // 1 bit
+} l1i_timing_mshr_entry;
+
+// We do not have access to the cache, so we aproximate it using this structure
+typedef struct __l1i_timing_cache_entry {
+  bool valid;              // 1 bit
+  uint64_t tag;            // L1I_TIMING_CACHE_TAG_BITS bits
+  uint64_t bere_line_addr; // 58 bits
+  bool accessed;           // 1 bit
+} l1i_timing_cache_entry;
+
+l1i_timing_mshr_entry l1i_timing_mshr_table[NUM_CPUS][L1I_TIMING_MSHR_SIZE];
+l1i_timing_cache_entry l1i_timing_cache_table[NUM_CPUS][L1I_SET][L1I_WAY];
+
+void l1i_init_timing_tables()
+{
+  for (uint32_t i = 0; i < L1I_TIMING_MSHR_SIZE; i++) {
+    l1i_timing_mshr_table[l1i_cpu_id][i].valid = 0;
+  }
+  for (uint32_t i = 0; i < L1I_SET; i++) {
+    for (uint32_t j = 0; j < L1I_WAY; j++) {
+      l1i_timing_cache_table[l1i_cpu_id][i][j].valid = 0;
+    }
+  }
+}
+
+uint64_t l1i_find_timing_mshr_entry(uint64_t line_addr)
+{
+  for (uint32_t i = 0; i < L1I_TIMING_MSHR_SIZE; i++) {
+    if (l1i_timing_mshr_table[l1i_cpu_id][i].tag == (line_addr & L1I_TIMING_MSHR_TAG_MASK) && l1i_timing_mshr_table[l1i_cpu_id][i].valid)
+      return i;
+  }
+  return L1I_TIMING_MSHR_SIZE;
```

Comment:
Yes, I think this is just asking for a `std::find_if`:

```suggestion
  auto it = std::find_if(std::begin(l1i_timing_mshr_table[l1i_cpu_id]), std::end(l1i_timing_mshr_table[l1i_cpu_id]), [tag=line_addr & L1I_TIMING_MSHR_TAG_MASK](const auto& x){ return x.valid && x.tag == tag; });
  return std::distance(std::begin(l1i_timing_mshr_table[l1i_cpu_id]), it);
```